community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -472,6 +474,98 @@ def update_topology(lkp: util.Lookup) -> None:
+    slurm_delete_reservation = "scontrol delete reservation {}"
+
+    reservation_name = node + "_maintenance"
+    ## If reservation exists for maintenance, update reservation if time window
+    ## for scheduled maintenance changes.
+    if reservation_name in list(reservation_map.keys()):
+      nodes, maintenance_start_time = reservation_map[reservation_name]
+      del reservation_map[reservation_name]  # remove the reservation from the map.
+
+      if maintenance_start_time == starttime:
+        return
+
+      util.run(slurm_delete_reservation.format(reservation_name), timeout=30)
+      util.run(slurm_create_reservation.format(starttime, nodes, reservation_name), timeout=30)
+
+    else: # Reservation doesn't exist, make reservation during maintenance window.
+      util.run(slurm_create_reservation.format(starttime, node, reservation_name), timeout=30)
+
+
+def get_slurm_reservation_maintenance() -> Dict[str, Tuple[str, str]]:
+    slurm_show_reservation = "scontrol show reservation --json"
+    res = util.run(slurm_show_reservation, timeout=30)
+    all_reservations = json.loads(res.stdout)
+    reservation_map = {}
+
+    for reservation in all_reservations['reservations']:
+        reservation_name = reservation['name']
+        nodes = reservation['node_list']
+        linux_time = reservation['start_time']['number']
+
+        if reservation_name is None or nodes is None or linux_time is None:
+          continue
+
+        # Check if the reservation is for scheduled maintenance.
+        maintenance_reservation = nodes + "_maintenance"
+        if reservation_name != maintenance_reservation:
+          continue
+
+        # Update maintenance start time compatible with slurm reservation format.
+        start_time = datetime.fromtimestamp(linux_time)
+        formatted_start_time = start_time.strftime('%Y-%m-%dT%H:%M:%S')
+        reservation_map[reservation_name] = (nodes, formatted_start_time)
+
+    return reservation_map
+
+
+def get_upcoming_maintenance(lkp: util.Lookup) -> Dict[str, str]:
+    upc_maint_map = {}
+
+    for node, properties in lkp.instances().items():
+        if 'upcomingMaintenance' in properties:
+          upc_maint_map[node] = properties['upcomingMaintenance']['startTimeWindow']['earliest']
+
+    return upc_maint_map
+
+# Sync maintenance reservation gets upcoming maintenance notification and
+# updates slurm reservation for the node during scheduled maintenance window.
+def sync_maintenance_reservation(lkp: util.Lookup) -> None:
+    upc_maint_map = get_upcoming_maintenance(lkp)
+    log.debug(f"upcoming-maintenance-vms: {upc_maint_map}")
+
+    # Get current slurm reservation for maintenance.
+    # [reservation_name --> nodes, start_time]
+    # reservation_name should be of format nodes_maintenance.
+    reservation_map = get_slurm_reservation_maintenance()
+    log.debug(f"reservation-map: {reservation_map}")
+
```

Comment:
**nit.** I would recommend to implement as:

```py
got = get_slurm_reservation_maintenance()  # map reservation_name -> time
want = ... # map resertvation_name -> (node_name, time)

to_remove = set(got.keys() - want.keys()) # set of reservation_names
to_create = {} # map reservation_name ->  (node_name, time)

for res_name, (node, time) in want.items():
  if res_name in got:
    if got[res_name] == time:
       # do nothing
    else:
       to_delete(res_name)
       to_create[res_name] = (node, time)
  else:
    to_create[res_name] = (node, time)

for res_name in to_delete:
  # ....
for res_name, (node, time) in to_create.items():  
  # ....
```

Reply:
This is much better approach. Thanks for suggesting.